### PR TITLE
[New Rule] AWS EC2 VPC Security Group Rule Added for Any Address or Remote Access Ports

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/ia-trade-team/issues/358

## Summary
This rule detects the addition of ingress rules to a VPC security group that allow traffic from any IP address (`0.0.0.0/0`) to sensitive ports commonly used for remote access, such as SSH (port 22) and RDP (port 3389). This configuration change can significantly increase the exposure of EC2 instances to potential threats, making it crucial to understand the context and legitimacy of such changes.

## Additional Information
- `0.0.0.0` allows inbound from any IPv4 address
- SSH and RDP are most common for remote connections, but we can add more

Screenshot 2024-05-09 at 5 34 09 PM

<details><summary>Example Data</summary>

```
{
  "_index": ".ds-logs-aws.cloudtrail-default-yyyy.mm.dd-000001",
  "_id": "unique_id_here",
  "_version": 1,
  "_score": 0,
  "_source": {
    "agent": {
      "name": "agent_name",
      "id": "agent_id",
      "type": "filebeat",
      "ephemeral_id": "ephemeral_id_here",
      "version": "agent_version"
    },
    "log": {
      "file": {
        "path": "https://bucket_name.s3.region.amazonaws.com/AWSLogs/account_id/CloudTrail/region/yyyy/mm/dd/log_file_name.json.gz"
      },
      "offset": log_offset
    },
    "elastic_agent": {
      "id": "agent_id",
      "version": "agent_version",
      "snapshot": false
    },
    "source": {
      "geo": {
        "continent_name": "Continent",
        "region_iso_code": "Region_Code",
        "city_name": "City",
        "country_iso_code": "Country_Code",
        "country_name": "Country",
        "region_name": "Region",
        "location": {
          "lon": longitude,
          "lat": latitude
        }
      },
      "as": {
        "number": as_number,
        "organization": {
          "name": "Org_Name"
        }
      },
      "address": "IP_Address",
      "ip": "IP_Address"
    },
    "tags": [
      "preserve_original_event",
      "forwarded",
      "aws-cloudtrail"
    ],
    "cloud": {
      "region": "aws_region",
      "account": {
        "id": "account_id"
      }
    },
    "input": {
      "type": "aws-s3"
    },
    "@timestamp": "timestamp_here",
    "ecs": {
      "version": "ecs_version"
    },
    "related": {
      "user": [
        "user_name"
      ]
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "aws.cloudtrail"
    },
    "tls": {
      "cipher": "cipher_suite",
      "client": {
        "server_name": "service.amazonaws.com"
      },
      "version": "tls_version",
      "version_protocol": "tls"
    },
    "aws": {
      "s3": {
        "bucket": {
          "name": "bucket_name",
          "arn": "bucket_arn"
        },
        "object": {
          "key": "object_key"
        }
      },
      "cloudtrail": {
        "event_version": "event_version",
        "flattened": {
          "request_parameters": {
            "fromPort": port_number,
            "groupId": "security_group_id",
            "toPort": port_number,
            "ipProtocol": "protocol",
            "cidrIp": "0.0.0.0/0"
          },
          "response_elements": {
            "_return": true,
            "requestId": "request_id",
            "securityGroupRuleSet": {
              "items": [
                {
                  "groupOwnerId": "account_id",
                  "fromPort": port_number,
                  "groupId": "security_group_id",
                  "isEgress": false,
                  "toPort": port_number,
                  "cidrIpv4": "0.0.0.0/0",
                  "ipProtocol": "protocol",
                  "securityGroupRuleId": "security_group_rule_id"
                }
              ]
            }
          }
        },
        "event_type": "AwsApiCall",
        "read_only": false,
        "user_identity": {
          "access_key_id": "access_key",
          "type": "IAMUser",
          "arn": "user_arn"
        },
        "recipient_account_id": "account_id",
        "event_category": "Management",
        "request_parameters": request_parameters_json,
        "request_id": "request_id",
        "response_elements": response_elements_json,
        "management_event": true
      }
    },
    "event": {
      "agent_id_status": "verified",
      "ingested": "ingestion_timestamp",
      "original": "original_log_json",
      "provider": "ec2.amazonaws.com",
      "created": "creation_timestamp",
      "kind": "event",
      "action": "AuthorizeSecurityGroupIngress",
      "id": "event_id",
      "type": [
        "info"
      ],
      "dataset": "aws.cloudtrail",
      "outcome": "success"
    },
    "user": {
      "name": "user_name",
      "id": "user_id"
    },
    "user_agent": {
      "original": "user_agent_string",
      "name": "user_agent_name",
      "device": {
        "name": "device_type"
      }
    }
  },
  "fields": {
    "aws.cloudtrail.request_parameters.text": [
      request_parameters_text
    ],
    "elastic_agent.version": [
      "agent_version"
    ],
    "tls.version_protocol": [
      "tls_protocol"
    ],
    "user_agent.original.text": [
      "user_agent_string"
    ],
    "aws.cloudtrail.flattened.response_elements": [
      flattened_response_elements
    ],
    "aws.cloudtrail.response_elements": [
      response_elements_text
    ],
    "agent.name.text": [
      "agent_name"
    ],
    "source.geo.region_name": [
      "region_name"
    ],
    "source.ip": [
      "IP_Address"
    ],
    "agent.name": [
      "agent_name"
    ],
    "event.agent_id_status": [
      "status"
    ],
    "source.geo.region_iso_code": [
      "Region_Code"
    ],
    "aws.cloudtrail.management_event": [
      true_or_false
    ],
    "event.kind": [
      "event_kind"
    ],
    "aws.cloudtrail.user_identity.arn": [
      "user_arn"
    ],
    "event.outcome": [
      "outcome"
    ],
    "source.geo.city_name": [
      "city_name"
    ],
    "tls.version": [
      "tls_version"
    ],
    "user_agent.original": [
      "user_agent_string"
    ],
    "event.original": [
      "original_event_json"
    ],
    "cloud.region": [
      "aws_region"
    ],
    "user.id": [
      "user_id"
    ],
    "input.type": [
      "input_type"
    ],
    "log.offset": [
      log_offset
    ],
    "user_agent.name": [
      "user_agent_name"
    ],
    "data_stream.type": [
      "stream_type"
    ],
    "related.user": [
      "related_user"
    ],
    "tags": [
      tags_array
    ],
    "event.provider": [
      "service_provider"
    ],
    "agent.id": [
      "agent_id"
    ],
    "ecs.version": [
      "ecs_version"
    ],
    "event.created": [
      "creation_timestamp"
    ],
    "aws.cloudtrail.event_version": [
      "event_version"
    ],
    "agent.version": [
      "agent_version"
    ],
    "source.as.number": [
      as_number
    ],
    "aws.cloudtrail.read_only": [
      true_or_false
    ],
    "aws.cloudtrail.event_category": [
      "event_category"
    ],
    "aws.cloudtrail.user_identity.type": [
      "user_type"
    ],
    "aws.s3.bucket.arn": [
      "bucket_arn"
    ],
    "aws.cloudtrail.recipient_account_id": [
      "account_id"
    ],
    "aws.cloudtrail.request_id": [
      "request_id"
    ],
    "tls.cipher": [
      "cipher_suite"
    ],
    "user.name": [
      "user_name"
    ],
    "source.geo.location": [
      {
        "coordinates": [
          longitude,
          latitude
        ],
        "type": "Point"
      }
    ],
    "source.address": [
      "IP_Address"
    ],
    "aws.cloudtrail.flattened.request_parameters": [
      {
        "fromPort": port_number,
        "groupId": "security_group_id",
        "toPort": port_number,
        "ipProtocol": "protocol",
        "cidrIp": "cidr_range"
      }
    ],
    "agent.type": [
      "agent_type"
    ],
    "event.module": [
      "aws"
    ],
    "source.geo.country_iso_code": [
      "Country_Code"
    ],
    "aws.cloudtrail.response_elements.text": [
      response_elements_text
    ],
    "elastic_agent.snapshot": [
      true_or_false
    ],
    "aws.cloudtrail.event_type": [
      "event_type"
    ],
    "aws.s3.bucket.name": [
      "bucket_name"
    ],
    "source.as.organization.name.text": [
      "Org_Name"
    ],
    "elastic_agent.id": [
      "agent_id"
    ],
    "data_stream.namespace": [
      "namespace"
    ],
    "source.as.organization.name": [
      "Org_Name"
    ],
    "source.geo.continent_name": [
      "Continent_Name"
    ],
    "tls.client.server_name": [
      "service_name"
    ],
    "event.action": [
      "event_action"
    ],
    "event.ingested": [
      "ingestion_timestamp"
    ],
    "@timestamp": [
      "event_timestamp"
    ],
    "cloud.account.id": [
      "account_id"
    ],
    "aws.cloudtrail.user_identity.access_key_id": [
      "access_key_id"
    ],
    "data_stream.dataset": [
      "dataset"
    ],
    "event.type": [
      "event_type"
    ],
    "log.file.path": [
      "log_file_path"
    ],
    "agent.ephemeral_id": [
      "ephemeral_id"
    ],
    "aws.cloudtrail.request_parameters": [
      request_parameters_json
    ],
    "event.id": [
      "event_id"
    ],
    "source.geo.country_name": [
      "Country_Name"
    ],
    "user_agent.device.name": [
      "device_name"
    ],
    "aws.s3.object.key": [
      "object_key"
    ],
    "event.dataset": [
      "dataset"
    ],
    "user.name.text": [
      "user_name"
    ]
  }
}

```

</details>

### CODE REVIEW

File: detection_rules/etc/non-ecs-schema.json [hunk 1]
```diff
@@ -144,5 +144,9 @@
     "auditd.data.a1": "keyword",
     "auditd.data.a2": "keyword",
     "auditd.data.a3": "keyword"
+  },
+  "logs-aws.cloudtrail-*": {
+    "aws.cloudtrail.flattened.request_parameters.cidrIp": "keyword",
+    "aws.cloudtrail.flattened.request_parameters.fromPort": "keyword"
```
Comment (brokensound77) [inline detection_rules/etc/non-ecs-schema.json:150]: any idea why these aren't populated in the integration?

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    "Use Case: Threat Detection",
+    "Tactic: Defense Evasion",
+]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.dataset: "aws.cloudtrail" and event.provider: ec2.amazonaws.com
+    and event.action: AuthorizeSecurityGroupIngress and event.outcome: success
+    and (aws.cloudtrail.flattened.request_parameters.cidrIp: 0.0.0.0/0 or aws.cloudtrail.flattened.request_parameters.fromPort: (22 or 3389))
```
Comment (brokensound77) [inline rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:85]:
```suggestion
event.dataset:"aws.cloudtrail" and event.provider:ec2.amazonaws.com
    and event.action:AuthorizeSecurityGroupIngress and event.outcome:success
    and (aws.cloudtrail.flattened.request_parameters.cidrIp:0.0.0.0/0 or aws.cloudtrail.flattened.request_parameters.fromPort:(22 or 3389))
```

Comment (brokensound77) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:85]: also do you know if `aws.cloudtrail.flattened.request_parameters.cidrIp:0.0.0.0/0` can be circumvented with `00.0.0.0/0` or similar variations, or does AWS normalize first

File: detection_rules/etc/non-ecs-schema.json (hunk 1 above)
Comment (terrancedejesus) [reply detection_rules/etc/non-ecs-schema.json:150]: The request and response dictionaries have important context that should be parsed out in general. Ref -> https://github.com/elastic/integrations/issues/9586

I am hoping to work with integrations to figure this out:

1. Do we need to give a list of nested structures to pull out if they exist?
2. Can we dynamically parse these into their own fields?

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml (hunk 2 above)
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml]: That is a great question. I would need to test this to confirm

**Review by Aegrah — APPROVED** [review]

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    "Use Case: Threat Detection",
+    "Tactic: Defense Evasion",
+]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.dataset:"aws.cloudtrail" and event.provider:ec2.amazonaws.com
+    and event.action:AuthorizeSecurityGroupIngress and event.outcome:success
+    and (aws.cloudtrail.flattened.request_parameters.cidrIp:0.0.0.0/0 or aws.cloudtrail.flattened.request_parameters.fromPort:(22 or 3389))
```
[the remark this answers is not in the document]
Comment (imays11) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:85]: +

File: detection_rules/etc/non-ecs-schema.json (hunk 1 above)
Comment (terrancedejesus) [reply detection_rules/etc/non-ecs-schema.json:150]: @brokensound77 - I do not anticipate that these changes to the integration will come any time soon, however, the issue is there for tracking as we can tune the rules when a decision and solution are determined. Any other thoughts before I resolve this?

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    Administrators may legitimately add security group rules to allow traffic from any IP address or from specific IP
+    addresses to common remote access ports.
+    """,
+]
+from = "now-60m"
+index = ["filebeat-*", "logs-aws.cloudtrail-*"]
+interval = "10m"
+language = "kuery"
+license = "Elastic License v2"
+name = "AWS EC2 VPC Security Group Ingress Rule Added for Any Address or Remote Access Ports"
```
[the remark this answers is not in the document]
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:27]: Good suggestion, renamed.

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+note ="""
+
+## Triage and Analysis
+
+### Investigating AWS EC2 VPC Security Group Ingress Rule Added for Any Address or Remote Access Ports
```
[the remark this answers is not in the document]
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:32]: Updated. Thanks!

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:85]: Added these ports.

[the remark this answers is not in the document]
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml:85]: Updated this and noticed I was missing ::/0 as well. Thanks!

File: rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml (hunk 2 above)
Comment (terrancedejesus) [reply rules/integrations/aws/defense_evasion_vpc_security_group_ingress_rule_added_for_remote_connections.toml]: Confirmed that AWS normalizes this.

**Comment by terrancedejesus** [comment]
@imays11 ready for your review when you have time :)

File: detection_rules/etc/non-ecs-schema.json (hunk 1 above)
Comment (brokensound77) [reply detection_rules/etc/non-ecs-schema.json:150]: Nope. Sgtm. Thanks for the details 👍

**Review by brokensound77 — APPROVED** [review]
🚀

**Pull request merged**